### PR TITLE
Add function and lambda accessor syntax

### PR DESCRIPTION
**Pull request by NikkyAI**
may be a partial duplicate of #63 but have this code anyhow.. maybe it helps even a little bit

this allows validations of java libraries without needing to write a `val ThatClass.someValue = getSomeValue()` extensions for each property

it also allows to use `aliased` custom accessors

a example out of our codebase using this to validate `com.nimbusds.jwt.SignedJWT`

```
fun ValidationBuilder<SignedJWT>.hasSubject(expectedSubject: String) {
    SignedJWT::getJWTClaimsSet required {
        JWTClaimsSet::getSubject required {
            addConstraint("subject must be '{0}'", expectedSubject) { subject ->
                subject == expectedSubject
            }
        }

    }
}

fun ValidationBuilder<SignedJWT>.hasClaimWithValue(claim: String, expected: String) {
    SignedJWT::getJWTClaimsSet required {
        val accessClaim = { it: JWTClaimsSet ->
            it.getStringClaim(claim) ?: null
        }
        accessClaim.required("getStringClaim($claim)") {
            addConstraint(
                "claim '{0}' must be '{1}'",
                claim,
                expected
            ) { value ->
                value == expected
            }
        }
    }
}
```

i added tests `functionAccessorSyntax`, `lambdaAccessorSyntax`, `complexLambdaAccessors` to demonstrate usage of this syntax

### CODE REVIEW

**Comment by dhoepelman** [comment]
If we wait long enough this will likely be solved by Kotlin itself, synthetic java accessors are scheduled for kotlin 2.1:
https://youtrack.jetbrains.com/issue/KT-8575/Support-Java-synthetic-property-references

File: src/commonTest/kotlin/io/konform/validation/ValidationBuilderTest.kt [hunk 1]
```diff
@@ -168,6 +170,95 @@ class ValidationBuilderTest {
...
+    }
+
+    @Test
+    fun lambdaAccessorSyntax() {
+        val splitDoubleValidation = Validation<Register> {
+            val getPassword = { r: Register -> r.password }
+            val getEmail = { r: Register -> r.email }
+            getPassword("getPasswordLambda") {
+                minLength(1)
+            }
```
Comment (dhoepelman) [inline src/commonTest/kotlin/io/konform/validation/ValidationBuilderTest.kt:200]: WDYT about this as an API for lambdas

```suggestion
            get("password", { it.password }) {
                minLength(1)
            }
```

I think that should be possible with
```kotlin
abstract fun <R> extract(name: String, prop: (T) -> R, init: ValidationBuilder<R>.() -> Unit): Constraint<T>
```

**Comment by dhoepelman** [comment]
Fixes #14

**Review by jillesvangurp — APPROVED** [review]

**Comment by NikkyAI** [comment]
yep, I think this works for us

**Comment by elijah-pl** [comment]
This issue you may find interesting 🙂
Having options for Kotlin pre-2.1 would be great though!
https://youtrack.jetbrains.com/issue/KT-8575/Support-Java-synthetic-property-references

**Comment by dhoepelman** [comment]
@NikkyAI @jillesvangurp Thanks a lot for the contribution! I've massaged the API into something I think can provide a good basis, see the new README. Intend to release this soon-ish as 0.7.0. Let me know if you have any feedback.

**Review by dhoepelman — APPROVED** [review]

**Comment by jillesvangurp** [comment]
Great!

**Pull request merged**